packages/backend/src/tokens/authenticateContext.ts

```diff
@@ -0,0 +1,88 @@
+import { constants } from '../constants';
+import type { ClerkRequest } from './clerkRequest';
+import type { AuthenticateRequestOptions } from './types';
+
+interface AuthenticateContextInterface extends AuthenticateRequestOptions {
+  // header-based values
+  sessionTokenInHeader: string | undefined;
+  origin: string | undefined;
+  host: string | undefined;
+  forwardedHost: string | undefined;
+  forwardedProto: string | undefined;
+  referrer: string | undefined;
+  userAgent: string | undefined;
+  secFetchDest: string | undefined;
+  accept: string | undefined;
+  // cookie-based values
+  sessionTokenInCookie: string | undefined;
+  clientUat: number;
+  // handshake-related values
+  devBrowserToken: string | undefined;
+  handshakeToken: string | undefined;
+  // url derived from headers
+  clerkUrl: URL;
+  // cookie or header session token
+  sessionToken: string | undefined;
+}
+
+interface AuthenticateContext extends AuthenticateContextInterface {}
+
+/**
+ * All data required to authenticate a request.
+ * This is the data we use to decide whether a request
+ * is in a signed in or signed out state or if we need
+ * to perform a handshake.
+ */
+class AuthenticateContext {
```

Comment:
This naming feels weird as it looks like a React Context. Do we even need a class at this point? This is a very nit-picky comment, so you can ignore it for now, but I wanted to raise the point that the abstraction might not help as much as intended.

Reply:
@nikosdouvlis Will we just pass a Request object afterall?

Reply:
BTW this is a great candidate for async local storage.

Reply:
> ... that the abstraction might not help as much as intended.
> BTW this is a great candidate for async local storage.

Let me reply to these two together, as my intention is for us to use `AuthenticateContext` or a similar struct with Async Local Storage, but only after we simplify type-handling in clerk/backend and we decide that the abstraction works for us

I will not refactor anything else to use `AuthenticateContext` as I've simplified the problematic types for now. Let's do a design session and talk more about it